packages/wallet/api/src/marshal-contexts.js

```diff
@@ -0,0 +1,362 @@
+// @ts-check
+import { makeScalarMap } from '@agoric/store';
+import { Far, makeMarshal, Remotable } from '@endo/marshal';
+import { HandledPromise } from '@endo/eventual-send'; // TODO: convince tsc this isn't needed
+
+const { details: X, quote: q } = assert;
+
+/**
+ * For a value with a known id in the board, we can use
+ * that board id as a slot to preserve identity when marshaling.
+ *
+ * @typedef {`board${Digits}`} BoardId
+ */
+
+/**
+ * @param {unknown} specimen
+ * @returns {specimen is BoardId}
+ */
+const isBoardId = specimen => {
+  return typeof specimen === 'string' && !!specimen.match(/board[^:]/);
+};
+
+/**
+ * When marshaling a purse, payment, etc. we partition the slots
+ * using prefixes.
+ *
+ * @template T
+ * @typedef {`${string & keyof T}:${Digits}`} WalletSlot<T>
+ */
+/**
+ * @template {string} K
+ * @typedef {`${K}:${Digits}`} KindSlot<K>
+ */
+
+/**
+ * @template T
+ * @param {T} _tables
+ * @param {string & keyof T} kind
```

Comment:
I think the above would also simplify this,
```suggestion
 * @param {keyof T} kind
```